extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/LoadedSegmentDataProvider.java

```diff
@@ -0,0 +1,234 @@
+
+  public LoadedSegmentDataProvider(
+      String dataSource,
+      ChannelCounters channelCounters,
+      ServiceClientFactory serviceClientFactory,
+      CoordinatorClient coordinatorClient,
+      ObjectMapper objectMapper,
+      QueryToolChestWarehouse warehouse
+  )
+  {
+    this.dataSource = dataSource;
+    this.channelCounters = channelCounters;
+    this.serviceClientFactory = serviceClientFactory;
+    this.coordinatorClient = coordinatorClient;
+    this.objectMapper = objectMapper;
+    this.warehouse = warehouse;
+  }
+
+  @VisibleForTesting
+  DataServerClient makeDataServerClient(FixedSetServiceLocator serviceLocator)
+  {
+    return new DataServerClient(serviceClientFactory, serviceLocator, objectMapper);
+  }
+
+  /**
+   * Queries a data server and returns a {@link Yielder} for the results, retrying if needed. If a dataserver indicates
+   * that the segment was not found, checks with the coordinator to see if the segment was handed off.
+   * - If the segment was handed off, returns with a {@link DataServerQueryStatus#HANDOFF} status.
+   * - If the segment was not handed off, retries with the known list of servers and throws an exception if the retry
+   * count is exceeded.
+   *
+   * @param <QueryType> result return type for the query from the data server
+   * @param <RowType> type of the result rows after parsing from QueryType object
+   */
+  public <RowType, QueryType> Pair<DataServerQueryStatus, Yielder<RowType>> fetchRowsFromDataServer(
+      Query<QueryType> query,
+      RichSegmentDescriptor segmentDescriptor,
+      Function<Sequence<QueryType>, Sequence<RowType>> mappingFunction,
+      Closer closer
+  ) throws IOException
+  {
+    final Query<QueryType> preparedQuery = Queries.withSpecificSegments(
+        query.withDataSource(new TableDataSource(dataSource)),
+        ImmutableList.of(segmentDescriptor)
+    );
+
+    Set<DruidServerMetadata> servers = segmentDescriptor.getServers();
+    DataServerClient dataServerClient = makeDataServerClient(FixedSetServiceLocator.forDruidServerMetadata(servers));
+    QueryToolChest<QueryType, Query<QueryType>> toolChest = warehouse.getToolChest(query);
+    Function<QueryType, QueryType> preComputeManipulatorFn =
+        toolChest.makePreComputeManipulatorFn(query, MetricManipulatorFns.deserializing());
+
+    final JavaType queryResultType = toolChest.getBaseResultType();
+    final int numRetriesOnMissingSegments = preparedQuery.context().getNumRetriesOnMissingSegments(DEFAULT_NUM_TRIES);
+
+    log.debug("Querying severs[%s] for segment[%s], retries:[%d]", servers, segmentDescriptor, numRetriesOnMissingSegments);
+    final ResponseContext responseContext = new DefaultResponseContext();
+
+    Pair<DataServerQueryStatus, Yielder<RowType>> statusSequencePair;
+    try {
+      statusSequencePair = RetryUtils.retry(
+          () -> {
+            Sequence<QueryType> sequence = dataServerClient.run(preparedQuery, responseContext, queryResultType, closer)
```

Comment:
What happen's if sequence is interrupted or closed?

Reply:
It would be thrown outside this class, which would cause the query to fail